Python/Auto_Caption_Generator/caption_generator_Xception.py

```diff
@@ -0,0 +1,468 @@
+from tqdm.notebook import tqdm
+import tensorflow as tf
+
+with tf.device('/gpu:0'):
+    config = tf.compat.v1.ConfigProto(gpu_options=
+                                      tf.compat.v1.GPUOptions(per_process_gpu_memory_fraction=1)
+                                      # device_count = {'GPU': 1}
+                                      )
+    config.gpu_options.allow_growth = True
+    session = tf.compat.v1.Session(config=config)
+    tf.compat.v1.keras.backend.set_session(session)
+    tqdm().pandas()
+    """
+    import tensorflow as tf
+    device_name = tf.test.gpu_device_name()
+    if device_name != '/device:GPU:0':
+      raise SystemError('GPU device not found')
+    print('Found GPU at: {}'.format(device_name))
+    """
+
+
+    def load_doc(filename):
+        """
+        Function for loading various files
+        :param filename: Source Path
+        """
+        # Opening the file as read only
+        file = open(filename, 'r')
+        text = file.read()
+        file.close()
+        return text
+
+
+    def all_img_captions(filename):
+        """
+        Load the captions of images
+        :param filename: Source path
+        :return: Captions
+        """
+        file = load_doc(filename)
+        captions = file.split('\n')
+        descriptions = {}
+        for caption in captions[:-1]:
+            img, caption = caption.split('\t')
+            if img[:-2] not in descriptions:
+                descriptions[img[:-2]] = [caption]
+            else:
+                descriptions[img[:-2]].append(caption)
+        return descriptions
+
+
+    def cleaning_text(captions):
+        """
+        Cleaning Captions, lower casing, removing puntuations and words containing numbers
+        :param captions:
+        :return: clean captions
+        """
+        table = str.maketrans('', '', string.punctuation)
+        for img, caps in captions.items():
+            for i, img_caption in enumerate(caps):
+                img_caption.replace("-", " ")
+                desc = img_caption.split()
+
+                # converts to lower case
+                desc = [word.lower() for word in desc]
+                # remove punctuation from each token
+                desc = [word.translate(table) for word in desc]
+                # remove hanging 's and a
+                desc = [word for word in desc if (len(word) > 1)]
+                # remove tokens with numbers in them
+                desc = [word for word in desc if (word.isalpha())]
+                # convert back to string
+
+                img_caption = ' '.join(desc)
+                captions[img][i] = img_caption
+        return captions
+
+
+    def text_vocabulary(descriptions):
+        """
+        Generate all the unique words from captions
+        :param descriptions: Captions
+        :return: Set of unique words
+        """
+        # build vocabulary of all unique words
+        vocab = set()
+        for key in descriptions.keys():
```

Comment:
Consider using `descriptions.items()` to iterate over the dictionary

Reply:
I don't understand the need for this one though, although it is a bit faster, I think the `keys` method is doing the best job here, I will make the changes though.